presto-main/src/main/java/com/facebook/presto/operator/HashBuilderOperator.java

```diff
@@ -538,35 +553,67 @@ private void unspillLookupSourceIfRequested()
         }
 
         verify(spiller.isPresent());
-        verify(!unspillInProgress.isPresent());
-
-        localUserMemoryContext.setBytes(getSpiller().getSpilledPagesInMemorySize() + index.getEstimatedSize().toBytes(), enforceBroadcastMemoryLimit);
-        unspillInProgress = Optional.of(getSpiller().getAllSpilledPages());
+        if (isSpillInSmallMemoryPool(operatorContext.getSession())) {
+            localUserMemoryContext.setBytes(index.getEstimatedSize().toBytes(), enforceBroadcastMemoryLimit);
+        }
+        else {
+            verify(!unspillInProgress.isPresent());
```

Comment:
It feels like we need to get rid of this codepath all together. Loading pages in a background thread is generally a very bad idea, as the CPU used for pages decoding is not accounted. Once we remove this codepath we would be able to get rid of the `ListenableFuture<List<Page>> getAllSpilledPages` interface and thread pools used in Spiller internally. Do you think we can try to gradually enable this codepath in production and remove it eventually?

CC: @rschlussel @highker